### PR TITLE
[dart][dart-dio] Nullable support/improvements

### LINKED ISSUE #5411: [BUG][dart][dart-dio] required fields are nullable in generated files
#### Bug Report Checklist

- [x] Have you provided a full/minimal spec to reproduce the issue?
- [x] Have you validated the input using an OpenAPI validator ([example](https://apidevtools.org/swagger-parser/online/))?
- [x] What's the version of OpenAPI Generator used?
- [x] Have you search for related issues/PRs?
- [x] What's the actual output vs expected output?
- [ ] [Optional] Bounty to sponsor the fix ([example](https://www.bountysource.com/issues/66123212-javascript-client-produces-a-wrong-object-for-a-string-enum-type-that-is-used-with-ref))

##### Description

When trying to generate dart-dio api client with open api specification with required fields, they are ignored by default.

I know that a nullableFields option is available for generation but there is only 2 possibles output:
- all fields are nullable
- all fields are non nullable

Why there is no way to have only required fields non nullable?

##### openapi-generator version

os: macOS
os version: 10.15.2
openapi-generator version: 4.2.3
node version: 10.15.2
npm version: 6.13.4

##### OpenAPI declaration file content or url

https://github.com/OpenAPITools/openapi-generator/blob/master/modules/openapi-generator/src/test/resources/3_0/petstore.yaml

##### Command line used for generation

```
npm install
npm run client:generate
```

with the following package.json
```
{
  "name": "dart-dio-gen",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "scripts": {
    "client:generate": "rm -rf client/ && openapi-generator generate --generator-name dart-dio --input-spec petstore.yaml --output client/"
  },
  "author": "",
  "license": "ISC",
  "devDependencies": {
    "@openapitools/openapi-generator-cli": "^1.0.10-4.2.3"
  }
}
```

##### Steps to reproduce

After generation, just have a look at `client/model/pet.dart` for example

Obtained generate file:

```
            import 'package:openapi/model/tag.dart';
            import 'package:built_collection/built_collection.dart';
            import 'package:openapi/model/category.dart';
        import 'package:built_value/built_value.dart';
import 'package:built_value/serializer.dart';

part 'pet.g.dart';

abstract class Pet implements Built<Pet, PetBuilder> {

    
        @nullable
    @BuiltValueField(wireName: r'id')
    int get id;
    
        @nullable
    @BuiltValueField(wireName: r'category')
    Category get category;
    
        @nullable
    @BuiltValueField(wireName: r'name')
    String get name;
    
        @nullable
    @BuiltValueField(wireName: r'photoUrls')
    BuiltList<String> get photoUrls;
    
        @nullable
    @BuiltValueField(wireName: r'tags')
    BuiltList<Tag> get tags;
    /* pet status in the store */
        @nullable
    @BuiltValueField(wireName: r'status')
    String get status;
        //enum statusEnum {  available,  pending,  sold,  };

    // Boilerplate code needed to wire-up generated code
    Pet._();

    factory Pet([updates(PetBuilder b)]) = _$Pet;
    static Serializer<Pet> get serializer => _$petSerializer;

}
```

Expected generated file:

```
            import 'package:openapi/model/tag.dart';
            import 'package:built_collection/built_collection.dart';
            import 'package:openapi/model/category.dart';
        import 'package:built_value/built_value.dart';
import 'package:built_value/serializer.dart';

part 'pet.g.dart';

abstract class Pet implements Built<Pet, PetBuilder> {

    
        @nullable
    @BuiltValueField(wireName: r'id')
    int get id;
    
        @nullable
    @BuiltValueField(wireName: r'category')
    Category get category;
    
    @BuiltValueField(wireName: r'name')
    String get name;
    
    @BuiltValueField(wireName: r'photoUrls')
    BuiltList<String> get photoUrls;
    
        @nullable
    @BuiltValueField(wireName: r'tags')
    BuiltList<Tag> get tags;
    /* pet status in the store */
        @nullable
    @BuiltValueField(wireName: r'status')
    String get status;
        //enum statusEnum {  available,  pending,  sold,  };

    // Boilerplate code needed to wire-up generated code
    Pet._();

    factory Pet([updates(PetBuilder b)]) = _$Pet;
    static Serializer<Pet> get serializer => _$petSerializer;

}
```

### PR DESCRIPTION
**Pull request by kuhnroyal**
Follow-up for #8235 and a requirement for any move towards Dart NNBD in #8278
Fixes #5411

### dart-dio/built_value

+ **Disable all fields nullable by default:**
This is not in line with OAS and will prevent future Dart nullabilty features (NNBD) from being useful as all types would be optional. This is a breaking change with fallback (set `nullableFields=true` ).

+ **Properties are `@nullable` when nullable OR ( not required AND no defaultValue )**

+ **Support nullable/required fields in JSON**
  * properties in built_value need to be `@nullable` when they are nullable in OAS or when they are not required in OAS
  * built_value does not support serializing `null` values by default as it is based on a serialization format based on iterables/lists and not maps
  * dart-dio uses the built_value json plugin to convert the built_value format to regular json
  * by generating a custom serializer for each class we can add support for serializing `null` values if the property is required AND nullable in OAS

### Dart
* basically just serialize `null` when the field is required and nullable

### PR checklist
 
- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] Pull Request title clearly describes the work in the pull request and Pull Request description provides details about how to validate the work. Missing information here may result in delayed response from the community.
- [x] Run the following to [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) and update samples:
  ```
  ./mvnw clean package 
  ./bin/generate-samples.sh
  ./bin/utils/export_docs_generators.sh
  ``` 
  Commit all changed files. 
  This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit as it would merge with master. 
  These must match the expectations made by your contribution. 
  You may regenerate an individual generator by passing the relevant config(s) as an argument to the script, for example `./bin/generate-samples.sh bin/configs/java*`. 
  For Windows users, please run the script in [Git BASH](https://gitforwindows.org/).
- [x] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master`, `5.1.x`, `6.0.x`
- [x] If your PR is targeting a particular programming language, @mention the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) members, so they are more likely to review the pull request.

CC @swipesight (2018/09) @jaumard (2018/09) @josh-burton (2019/12) @amondnet (2019/12) @sbu-WBT (2020/12) @kuhnroyal (2020/12) @agilob (2020/12)

### CODE REVIEW

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/DartDioClientCodegen.java [hunk 1]
```diff
@@ -40,7 +40,7 @@
 
     private static final String CLIENT_NAME = "clientName";
 
-    private boolean nullableFields = true;
```
Comment (agilob) [inline modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/DartDioClientCodegen.java:43]: What do you think to move this to dart generator?

**Comment by agilob** [comment]
@wing328  is there any other generator that supports not-null-by-default?

@kuhnroyal  Most languages are null-by-default, can this change potentially make dart-dio incompatible wither other generators without changes to spec file? I thought adding NNBD would be a bit simpler to start with, by just changing
 `int val` into `int? val` if no default value is specified. What's your plan for next changes?

**Comment by kuhnroyal** [comment]
@agilob Kotlin is NNBD and basically the same as Dart in this regard. I have looked at the templates there and basically mirrored this with some adaptions for built_value.

Basically everything that is nullable in OAS needs to be an optional type in Dart. Everything else should be non-null. Doing this with the current mutable models in the dart2 generator isn't great but possible, it should be straight forward for dart-dio.

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/DartDioClientCodegen.java (hunk 1 above)
Comment (kuhnroyal) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/DartDioClientCodegen.java:43]: I would rather deprecate this and get rid of it in 6.0 -  we should not promote this going forward in with Dart/NNBD. I think this was more of a workaround for built_value. Maybe @josh-burton can say more to the original motivation.

**Comment by kuhnroyal** [comment]
I would add a generator NNBD switch that defaults to true for 5.1.0 and allow users to set this to false.

**Review by agilob — APPROVED** [review]

**Comment by wing328** [comment]
> @wing328 is there any other generator that supports not-null-by-default?

`csharp-netcore` client generator comes to my mind

**Pull request merged**